docs: Rack awareness in Kubernetes (#946)
redpanda-data/documentation: modules/manage/pages/kubernetes/k-manage-resources.adoc

[[memory]]
== Configure memory resources

> On a worker node, Kubernetes and Redpanda processes are running at the same time, including the Seastar subsystem that is built into the Redpanda binary. Each of these processes consumes memory. Proper configuration of memory resources ensures optimal performance and stability for your Redpanda deployment.

Reviewer: I don't think it makes sense to talk about Seastar as a separate subsystem. It's a framework that redpanda is built in and doesn't really do anything besides running redpanda. Do we discuss it like this elsewhere in our docs?

> === Memory allocation and Seastar flags

Reviewer: OOC do we talk about resource allocation on baremetal? Last I did a search the only instances of `--memory` and the like showed up in our Kubernetes docs which should ideally be abstracting those details away 😅

Author: We have a Jira to document the relevant Seastar flags for bare-metal deployments. But since we don't have that yet, I wanted to minimally explain the two flags that affect the two memory management options in K8s.

Reviewer: Ah gotcha. In that case I'm going to defer to either @dotnwat or @travisdowns on the descriptions here to avoid saying anything incorrect. My mental model of `--reserve-memory` is that it's needlessly convoluted which we probably don't want to put in our docs.

Reviewer 2: @JakeSCahill @chrisseto can I get some background/context on this doc? Why do we document the `--memory` and `--reserve-memory` flags here? If anything they could be documented in a non-k8s part of the doc, or maybe we don't have that.

I'm happy to give precise definitions of these flags (I don't think the below is right), but let's first decide if we want to document them and in what context.

Author: > We have a Jira to document the relevant Seastar flags for bare-metal deployments. But since we don't have that yet, I wanted to minimally explain the two flags that affect the two memory management options in K8s.

@travisdowns, we're documenting the flags here because we don't yet have docs for them ([ticket is here](https://redpandadata.atlassian.net/browse/DOC-178)). Since we have two ways of setting memory resources in Kubernetes, and the differences are in how we configure those flags, I thought it would be useful to explain them to users.

We could gloss over the differences and say that one is legacy and not for production, but for completeness, it might be better to explain the reasons.

Reviewer 2: @JakeSCahill sure makes sense. These flags are bit hard to give independent definitions for because the behavior of one depends on whether the other is set, so there's really a 2x2 matrix of {not set, set} for each flag we need to consider.

This is the write-up of the behavior of these two flags I refer to and try to keep up to date:

https://github.com/scylladb/seastar/issues/375#issuecomment-2530012089

I just now added the 2x2 matrix explicitly. Does that give enough detail? Does it work for your documentation purposes?

Author: That's great. I've added some of this extra info to a collapsible block for those who want the details.

Redpanda uses the following Seastar flags to control memory allocation:

[cols="1m,2a"]
|===

> |Seastar Flag|Description

Reviewer: Should we mention `--smp` here as well?

Author: This part is just explaining how to allocate memory resources to Redpanda.

.Learn more about these Seastar flags
[%collapsible]
====

[cols="1,1,2"]
|===
|                        | **`--memory` Not set**
|                       **`--memory` Set (M)**

| **`--reserve-memory` Not set**  
| Heap size = available memory - calculated reserve
| Heap size = exactly M (if M plus calculated reserve ≤ available memory). Otherwise, startup fails

| **`--reserve-memory` set \(R)**
| Heap size = available memory - R
| Heap size = exactly M (if M + R ≤ available memory). Otherwise, startup fails
|===

Definitions and behavior:

- **Available memory**: The memory remaining after subtracting system requirements, such as `/proc/sys/vm/min_free_kbytes`, from the total or cgroup-limited memory.
- **Calculated reserve**: The greater of 1.5 GiB or 7% of _available memory_ used when `--reserve-memory` is not explicitly set.

====

**Legacy (default) behavior**: By default, the Helm chart and the Redpanda CRD define memory requirements using the `resources.memory.container` configuration, which allocates:

- 80% of the container memory to `--memory`
- 20% to `--reserve-memory`

> CAUTION: This default legacy behavior is for backward compatibility only. **Do not use it in production.**

Reviewer: We already have people using this in production and it's the default. It's not that it's bad or unstable, the interface just doesn't make sense and has limited customization potential on the Kubernetes side.

How about?
```suggestion
This default legacy behavior is for backward compatibility. It is deprecated as the model it presents is difficult to understand and doesn't allow sufficient control of the redpanda container's resource requirements.
```

> **Production recommendation**: Use `resources.requests.memory` for production deployments to ensure predictable memory allocation by letting Kubernetes schedule and enforce resource limits. When you set a value for `resources.requests.memory`:

Reviewer: Requests aren't actually enforced by kubelet, they're just used to binpack containers onto hosts. Limits are enforced. 

And to nit pick on a specific detail here, If `resources.requests.memory` is not set but `resources.limits.memory` is:
```yaml
resources:
  requests: {} # Neither memory or cpu specified
  limits:
    memory: 9001Gi # Only a limit
```

the chart will fallback to using the limits, just like Kubernetes' scheduler does.

- Kubernetes allocates 90% of the requested memory to the Redpanda heap (using `--memory`).
+

> The remaining 10% of the requested memory is reserved for container overhead. This allocation ensures that system processes and runtime requirements have sufficient memory while providing a predictable, dedicated heap for Redpanda.

Reviewer: So this is where my knowledge starts to get a bit fuzzy. If Travis has said otherwise, take his word over mine.

There's no "container overhead". It's more of a fudge factor in the context of a container.

Outside of a containerized context, reserve memory accounts for everything else running on the machine. Primarily the OS itself.

In a containerized context, that reservation is assumed to be accounted for by the cgroup hierarchy. For example, kubelet slices off a good chunk of CPU and memory for itself. What we need to account for instead is:
- imperfect memory account. It's (reportedly) possible to exceed it's memory bounds slightly. If a container goes half a kb over it's configured limit, it might be OOMKilled. We don't want that.
- Liveness and readiness probes. We exec `rpk` for our probes. Any memory it uses counts towards the containers total memory utilization.
- `kubectl exec`. Debugging happens a lot, just like the probes any additional processes in the container count towards the total memory usage.
- `emptyDir` mount, weirdly, count towards memory usage.

I may just be picking on the term container overhead here. How does

> The remaining 10% of the memory is reserved for non-redpanda uses  that count towards the container's total memory usage, such as exec probes, [`emptyDir`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#memory-backed-emptydir)s, and `kubectl exec`.  This ensures occational bursts of memory use don't result in redpanda being OOMKilled.

sound to you?

> - The `--reserve-memory` value is fixed at 0 because Kubernetes already manages container overhead.

Reviewer: Still picking on container overhead here. Though I think it would be good to distinguish the intended purpose of the `--reserve-memory` flag and the 10% overhead we carve out.

```suggestion
- The `--reserve-memory` value is fixed at 0 because Kubelet already manages system level memory reservations.
```

    resources:
      memory:
        enable_memory_locking: true <1>

>       requests:

Reviewer: `resources.{requests,limits}` is mutually exclusive with `resources.{memory,cpu,container}`. If requests _and_ limits are provided, memory, cpu, and container, will be ignored.

Memory locking can be enabled with either `statefulset.additionalRedpandaCmdFlags` or some values in `config`. I know the key is `enable_memory_locking` but I'm not sure which stanza it goes into.

Author: Our existing doc already mentioned `resources.memory.enable_memory_locking`. Should we keep that as a recommended way of setting memory locking? Or is it better to use `statefulset.additionalRedpandaCmdFlags`?

Reviewer: Do we have any baremetal docs that mention memory locking? If so, I'd vote to align with those be it the CLI flag or the rpk config.

If not, I guess it's dealers choice? I don't like `resources.memory.enable_memory_locking` because it's specific to kubernetes. I like to keep interfaces as consistent as possible so transitioning from baremetal to k8s or visa versa will lead to users thinking: "`<blank>` is a CLI flag/RPK Config. Okay, how do I set that in Kubernetes/Baremetal" rather than needing to have a bespoke answer for every piece of configuration.

Author: Updated to use `additionalRedpandaCmdFlags`

```bash
helm upgrade --install redpanda redpanda/redpanda --namespace <namespace> --create-namespace \
  --set resources.memory.enable_memory_locking=true \ <1>
  --set resources.requests.memory=<number><unit> <2>
```

====
--
======

<1> Enabling memory locking prevents the operating system from paging out Redpanda's memory to disk. This can significantly improve performance by ensuring Redpanda has uninterrupted access to its allocated memory.

> <2> Allocate at least 2.5 Gi of memory per core to ensure Redpanda has the 2 Gi per core it requires after accounting for the 90% allocation to the `--memory` flag. Redpanda supports the following memory resource units: B, K, M, G, Ki, Mi, and Gi. Memory units are converted to the nearest whole MiB. For a description of memory resource units, see the https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory[Kubernetes documentation^].

Reviewer: > Memory units are converted to the nearest whole MiB.

nit: I'm not 100% certain how to best phrase this because I'd describe it in code 😓 but memory units are `floor`ed or truncated to MiB. Rounded down, maybe?

> Allocate at least 2.5 Gi of memory per core to ensure Redpanda has the 2 Gi per core it requires after accounting for the 90% allocation to the `--memory` flag.

This math doesn't add up. Wouldn't that be ~2.2Gi per core? Or do we want there to be 2.5Gi per core intentionally including the overhead? (This may be a good question for core performance)

Author: Updated. This was a left-over from when the default was 80%

If Redpanda runs in a shared environment, where multiple applications run on the same worker node, you can make Redpanda less aggressive in CPU usage by enabling overprovisioning. This adjustment ensures a fairer distribution of CPU time among all processes, improving overall system efficiency at the cost of Redpanda's performance.

> You can enable overprovisioning by either setting the CPU request to a fractional value or setting `resources.cpu.overprovisioned` to `true`.

Reviewer: nit: clearify that fractional must be < 1 for overprovisioned to be set
```suggestion
You can enable overprovisioning by either setting the CPU request to a fractional value less than 1 or setting `resources.cpu.overprovisioned` to `true`.
```

And similarly to the comment above, overprovisioned can be set with either a CLI flag or an RPK config as well. Let's align on either CLI flags or RPK config values for this as well.

> NOTE: When `resources.requests` or `resources.limits` are set, the `resources.cpu` parameter (including cores) is ignored. Ensure that you have not configured CPU requests and limits explicitly to avoid unexpected behavior in shared environments.

Reviewer: This condition is that both requests and limits are set:
```suggestion
NOTE: When `resources.requests` and `resources.limits` are set, the `resources.cpu` parameter (including cores) is ignored. Ensure that you have not configured CPU requests and limits explicitly to avoid unexpected behavior in shared environments.
```

I'm not sure what the final sentence here means?

You can enable overprovisioning by either setting the CPU request to a fractional value or setting `overprovisioned` to `true`.

> NOTE: Setting `resources.requests.cpu` to a fractional value, such as 200m, enables Kubernetes to schedule Pods alongside other workloads efficiently, ensuring fair resource distribution. However, this may impact Redpanda's performance under heavy loads.

Reviewer: This is where things start to get a bit nuanced. redpanda can't effectively make use of fractional CPU values. So the `--smp` flag is always rounded down with a minimum value of one. So 200m (--smp=1) will only work with the `overprovisioned` flag and `2500m` (--smp=2) will only result in redpanda utilizing 2 cores.

Author: is this true for `resources.cpu.cores`? Our docs currently say that fractional CPU values are allowed for that config.

I think given this:

> resources.{requests,limits} is mutually exclusive with resources.{memory,cpu,container}. If requests and limits are provided, memory, cpu, and container, will be ignored.

I should revert changes to the shared environments section since users will still need to use `resources.cpu` settings. Perhaps we need a note like:

```asciidoc
NOTE: When `resources.requests` or `resources.limits` are set,
the `resources.cpu` parameter (including cores) is ignored.
Ensure that you have not configured CPU requests and limits explicitly
to avoid unexpected behavior in shared environments.
```

If Redpanda runs in a shared environment, where multiple applications run on the same worker node, you can make Redpanda less aggressive in CPU usage by enabling overprovisioning. This adjustment ensures a fairer distribution of CPU time among all processes, improving overall system efficiency at the cost of Redpanda's performance.

You can enable overprovisioning by either setting the CPU request to a fractional value less than 1 or enabling the `--overprovisioned` flag.

> NOTE: You cannot enable overprovisioning when both `resources.requests` and `resources.limits` are set. When both of these configurations are set, the `resources.cpu` parameter (including cores) is ignored.

Reviewer: Good call out but let's clarify that this is specific to the parameter and not the overprovisioned flag in general.

```suggestion
NOTE: You cannot enable overprovisioning via `resources.cpu.overprovisioned` when both `resources.requests` and `resources.limits` are set. When both of these configurations are set, the `resources.cpu` parameter (including overprovisioned and cores) is ignored.
```

+
```bash
helm upgrade --install redpanda redpanda/redpanda --namespace <namespace> --create-namespace \
  --set resources.requests.cpu=<number-of-cpu-cores> \

>   --set resources.cpu.overprovisioned=true

Reviewer: Ah, so same deal with the overprovisioned flag. It's now a CLI flag (like it is in redpanda) or it can be set in the redpanda.yaml which causes rpk to set the flag on you behalf.
